### ISSUE TITLE
Something went wrong since last commit

### ISSUE THREAD

**Issue by AlessandroZ** [n1nj4sec/pupy#358]
I have sync my fork to the last commit and downloaded new binaries. However, I do not manage to get a reverse shell (I have tested it on Linux and Windows). 

When I generate a linux binary, the binary is created but I have two lines in red (Required credentials and the line with "SSL_BIND_CERT, SSL_CA_CERT"). And no connection is made when I launch the binary: 
```
./pupygen.py -f client -A x64 -O linux connect --host 192.168.142.163:1234
[+] Generate client: linux/x64
[C] launcher: connect
[C] launcher_args: ['--host', '192.168.142.163:1234']
[C] offline_script: 
[+] Required credentials:
[+] SSL_BIND_CERT, SSL_CA_CERT, SSL_CLIENT_CERT, SSL_BIND_KEY, SSL_CLIENT_KEY
[+] OUTPUT_PATH = /tmp/pupyx64.mIsHAJ.lin
[+] SCRIPTLETS = []
[+] DEBUG = False
```

Moreover, when I generate a python payload, I have some errors: 
```
sudo ./pupygen.py -f py connect --host 192.168.142.163:1234
WARNING:root:Error with GnomeKeyring get_pass : Did not receive a reply. Possible causes include: the remote application did not send a reply, the message bus security policy blocked the reply, the reply timeout expired, or the network connection was broken.
[I] use_gnome_keyring is true, the password will be stored in the Gnome-Keyring
[I] Credentials password: 
WARNING:root:Error with GnomeKeyring store_pass : Did not receive a reply. Possible causes include: the remote application did not send a reply, the message bus security policy blocked the reply, the reply timeout expired, or the network connection was broken.
ERROR:root:bad decrypt pad (123)
Traceback (most recent call last):
  File "./pupygen.py", line 545, in <module>
    pupygen(parser.parse_args(), config)
  File "./pupygen.py", line 483, in pupygen
    packed_payload=pack_py_payload(get_raw_conf(conf))
  File "./pupygen.py", line 73, in get_raw_conf
    credentials = Credentials(role='client')
  File "/tmp/pupy/pupy/pupylib/PupyCredentials.py", line 206, in __init__
    encryptor.decrypt(StringIO(content), fcontent)
  File "/tmp/pupy/pupy/pupylib/PupyCredentials.py", line 162, in decrypt
    raise ValueError("bad decrypt pad (%d)" % padding_length)
ValueError: bad decrypt pad (123)
bad decrypt pad (123)
```
I have tested to clone the project from @alxchk (and binaries too) and same problem with the connection (I didn't manage to get a reverse shell working).
I could generate the python payload without errors this time but it didn't work. 

I don't understand as well, what is stored on the gnome keyring. I saw that a credentials password is required (at least on the @alxchk fork) but I didn't manage to reset it. 

So I don't know if something has been broken or if it's my pupy installation.

**Comment by alxchk** [comment]
Try to decrypt your credentials manually. Looks like pupysh have some issues with communication with gnome-keyring and cant get password from it. 

Currently file with credentials is encrypted using password. Also there may be situation, when pupysh and pupygen have different ideas about where credentials are stored. Try to check both locations: 
~/.config/pupy/credentials.py and crypto/credentials.py in your pupy or cwd

**Comment by AlessandroZ** [comment]
I have an encrypted file here "pupy/crypto/credentials.py"
```
file pupy/crypto/credentials.py 
pupy/crypto/credentials.py: data
```
But I don't have a pupy directory on my ".config" directory (I have checked on my regular user and on the root directory). It could be the problem. 

To understand a little bit, when you talked about credentials, which ones are stored on the credentials.py file. Is it users credentials from "data/db/creds.json" or credentials used by pupy ?

**Comment by alxchk** [comment]
creds.json - for credentials extracted from machines
credentials.py - for credentials used by pupy

Logic to select location for credentials.py placement defined in https://github.com/alxchk/pupy/blob/unstable/pupy/pupylib/PupyCredentials.py#L182 and https://github.com/alxchk/pupy/blob/unstable/pupy/pupylib/PupyConfig.py#L122

TLDR - it tries to get file either from pupy crypto/credentials.py (exactly where you found your file) or home/project directory.

It's fine to have credentials.py in your working/pupy directory, so this shouldn't be an issue. You can check that you know right password using this command.

```
openssl enc -d -aes-256-cbc -md md5 -in credentials.py -out credentials.dec.py
```

I think your issue because of using sudo + gnome-keyring. Likely you launch sudo to use privileged port, and pupy cant communicate with gnome keyring daemon after that. Try without sudo first. Also you can try to build payload with --debug, and launch pupysh with debug also.

**Comment by AlessandroZ** [comment]
No, same issue without sudo or launching everything with the root user. I still not managed to get a working reverse shell. 

I don't know how to build a debug template so I get this error (output from pupygen.py and --debug option): 
`ValueError: Template not found (/tmp/pupy/pupy/payload_templates/pupyx64d.lin)`

I don't know if it's because of Gnome keyring but lots of times (not always), pupysh is quite long to start (almost as long as metasploit lol).

**Comment by AlessandroZ** [comment]
I don't know if using a system wallet it's a good idea, because it depends to much on the system (someone on KDE could not manage to use it). 

Even if these files are sensitive, I don't know if it's necessary to encrypt it. I think it's an heavy process which could be quite painful. 

Maybe we could add an option on a configuration file to use it or not.

**Comment by alxchk** [comment]
```
ValueError: Template not found (/tmp/pupy/pupy/payload_templates/pupyx64d.lin)
```
This means that you don't have debug template built. You can build python-only payload with something like
```
./pupygen.py -f py -o /tmp/test.py --debug connect --host 127.0.0.1:1234 -t ssl
```

> I don't know if it's because of Gnome keyring but lots of times (not always), pupysh is quite long to start (almost as long as metasploit lol).
> Maybe we could add an option on a configuration file to use it or not.

You can try to check that all external stuff is disabled. By default almost all  is disabled

```
[pupyd]
igd = false
use_gnome_keyring = false
allow_requests_to_external_services = false
```

Also you can check actual config from pupysh (BTW you can generate payloads from pupysh also):
```
config list pupyd
```


If you want to build templates manually, use

```
cd PUPYDIR/clients/sources-linux
./buildenv.sh
./build.sh
```

To save your time you can try first with build from CI where all is included already. Take one from here:
https://github.com/alxchk/pupy/releases

**Comment by alxchk** [comment]
> Even if these files are sensitive, I don't know if it's necessary to encrypt it. I think it's an heavy process which could be quite painful.

This is optional, but enabled by default. If you start pupysh with -NE option it will not encrypt newly generated credentials. Or you can simply decrypt them. If creds are not encrypted, then they will not be reencrypted.

**Comment by alxchk** [comment]
> To save your time you can try first with build from CI where all is included already. Take one from here:
https://github.com/alxchk/pupy/releases

Well, it turns out that there is a bug in txz generation. Ignore this for an hour >_<

**Comment by AlessandroZ** [comment]
Ok so, I have removed the credentials.py file to use a non encrypted file, I launched pupysh with the -NE option and it has created a new credentials.py file not encrypted. 

Creating a python payload as you told me works well. However, I have tried to generate windows and Linux payloads and no one worked. 
Maybe it could be a problem on the templates. 

I will try to investigate when I will have more time.

**Comment by alxchk** [comment]
> However, I have tried to generate windows and Linux payloads and no one worked.

If it fails with the error line you post here last time (Template not found) this simply mean that templates are not build.  You can either get them from @n1nj4sec pupy-binaries repo, or extract from CI archives. I suggest to use CI's builds or build manually, because now quite a lot of things were done in pp/pupyimporter which requires rebuilds. Also if you try my repo, try unstable branch. It's more recent and stable than master :D

**Comment by AlessandroZ** [comment]
Ok ok, my fault !!!

I have mixed older payloads and the new ones. Trying to resolve the Gnome keyring problem, I have tried to use different templates and when this problem has been fixed I didn't replace the good template with the last one. Sorry, now it works. I have tried windows and linux binaries and it works. 

Thanks for your time, at least I understood the goal of Gnome keyring on pupy :) 

I close the issue.

**Comment by n1nj4sec** [comment]
I added the gnome-keyring feature to avoid retyping my credentials every time and still have the credentials encrypted ! I know the creds are easily retrievable with laZagne but it's still better than cleartext ;)

**Comment by AlessandroZ** [comment]
I think, it could be better to add the gnome keyring functionality as false by default and not [True](https://github.com/n1nj4sec/pupy/blob/master/pupy/pupy.conf.default#L24) or adding some explaination on the Wiki. 
Otherwise, I think many people will have the same issue as me and won't be able to understand the problem.

**Comment by n1nj4sec** [comment]
yup, agreed
